lib/onc_certification_g10_test_kit/smart_invalid_pkce_group.rb

```diff
@@ -0,0 +1,244 @@
+module ONCCertificationG10TestKit
+  class InvalidSMARTTokenRequestTest < Inferno::Test
+    title ' OAuth token exchange fails when supplied invalid code_verifier'
+    description %(
+      If the request failed verification or is invalid, the authorization
+      server returns an error response.
+    )
+    uses_request :redirect
+    id :invalid_pkce_request
+
+    input :code, :use_pkce, :pkce_code_verifier, :client_id, :client_secret, :smart_token_url
+
+    def modify_oauth_params(oauth_params)
+      oauth_params
+    end
+
+    run do
+      skip_if request.query_parameters['error'].present?, 'Error during authorization request'
+
+      oauth2_params = {
+        grant_type: 'authorization_code',
+        code: code,
+        redirect_uri: config.options[:redirect_uri]
+      }
+
+      oauth2_headers = { 'Content-Type' => 'application/x-www-form-urlencoded' }
+
+      if client_secret.present?
+        client_credentials = "#{client_id}:#{client_secret}"
+        oauth2_headers['Authorization'] = "Basic #{Base64.strict_encode64(client_credentials)}"
+      else
+        oauth2_params[:client_id] = client_id
+      end
+
+      modify_oauth_params(oauth2_params)
+
+      post(smart_token_url, body: oauth2_params, name: :token, headers: oauth2_headers)
+
+      assert_response_status(400)
```

Comment:
Are we sure this has to be a 400?  Should we instead just ensure it doesn't succeed (e.g. a 20x and there is no response body with the bearer token in it?).  Is the language in the spec strong enough to dictate exactly how the system reports a problem back to the client?

Reply:
Hmm, I think 401 may also be acceptable for these particular cases.

Reply:
Ok, I feel like we always continually back off the specific status code, but 400,401 is good enough for now :)